tools/common/Assembly.cs

```diff
@@ -130,132 +228,145 @@ public void ExtractNativeLinkInfo ()
+			if (metadata.ForceLoad && !(metadata.SmartLink && App.Registrar == RegistrarMode.Static))
+				ForceLoad = true;
 
-						Frameworks.Add (path);
-					} else {
-						if (!Application.IsUptodate (FullPath, path)) {
-							Application.ExtractResource (assembly.MainModule, libraryName, path, false);
-							Driver.Log (3, "Extracted third-party binding '{0}' from '{1}' to '{2}'", libraryName, FullPath, path);
-							LogLinkWithAttribute (linkWith);
-						} else {
-							Driver.Log (3, "Target '{0}' is up-to-date.", path);
-						}
+			if (!string.IsNullOrEmpty (metadata.LinkerFlags)) {
+				if (LinkerFlags == null)
+					LinkerFlags = new List<string> ();
+				LinkerFlags.Add (metadata.LinkerFlags);
+			}
 
-						if (!File.Exists (path))
-							ErrorHelper.Warning (1302, "Could not extract the native library '{0}' from '{1}'. " +
-							"Please ensure the native library was properly embedded in the managed assembly " +
-							"(if the assembly was built using a binding project, the native library must be included in the project, and its Build Action must be 'ObjcBindingNativeLibrary').",
-								libraryName, path);
+			if (!string.IsNullOrEmpty (metadata.Frameworks)) {
+				foreach (var f in metadata.Frameworks.Split (new char [] { ' ' })) {
+					if (Frameworks == null)
+						Frameworks = new HashSet<string> ();
+					Frameworks.Add (f);
+				}
+			}
 
-						LinkWith.Add (path);
-					}
+			if (!string.IsNullOrEmpty (metadata.WeakFrameworks)) {
+				foreach (var f in metadata.WeakFrameworks.Split (new char [] { ' ' })) {
+					if (WeakFrameworks == null)
+						WeakFrameworks = new HashSet<string> ();
+					WeakFrameworks.Add (f);
 				}
 			}
 
-			if (exceptions != null && exceptions.Count > 0)
-				throw new AggregateException (exceptions);
+			if (metadata.NeedsGccExceptionHandling)
+				NeedsGccExceptionHandling = true;
 
-			// Make sure there are no duplicates between frameworks and weak frameworks.
-			// Keep the weak ones.
-			if (Frameworks != null && WeakFrameworks != null)
-				Frameworks.ExceptWith (WeakFrameworks);
+			if (metadata.IsCxx)
+				EnableCxx = true;
 
-			if (NeedsGccExceptionHandling) {
-				if (LinkerFlags == null)
-					LinkerFlags = new List<string> ();
-				LinkerFlags.Add ("-lgcc_eh");
+#if MONOTOUCH
+			if (metadata.Dlsym != DlsymOption.Default)
+				App.SetDlsymOption (FullPath, metadata.Dlsym == DlsymOption.Required);
+#endif
+		}
+
+		string ExtractNativeLibrary (AssemblyDefinition assembly, NativeReferenceMetadata metadata)
+		{
+			string path = Path.Combine (App.Cache.Location, metadata.LibraryName);
+
+			if (!Application.IsUptodate (FullPath, path)) {
+				Application.ExtractResource (assembly.MainModule, metadata.LibraryName, path, false);
+				Driver.Log (3, "Extracted third-party binding '{0}' from '{1}' to '{2}'", metadata.LibraryName, FullPath, path);
+				LogLinkWithAttribute (metadata.Attribute);
```

Comment:
`LogLinkWithAttribute` should be changed to log the metadata instead (and called when the metadata comes from the manifest as well). It would also be nice to have logging that stated where the native resources came from (linkwith vs manifest).